opm/parser/eclipse/Parser/Parser.cpp

```diff
@@ -42,60 +42,194 @@
+        if( input.empty() ) return false;
+
+        auto end = std::find( input.begin(), input.end(), '\n' );
+
+        line = string_view( input.begin(), end );
+        const auto mod = end == input.end() ? 0 : 1;
+        input = string_view( end + mod, input.end() );
+        return true;
+    }
+
+    /*
+     * Read the input file and remove everything that isn't interesting data,
+     * including stripping comments, removing leading/trailing whitespaces and
+     * everything after (terminating) slashes
+     */
+    static inline std::string clean( const std::string& str ) {
+        std::string dst;
+        dst.reserve( str.size() );
+
+        string_view input( str ), line;
+        while( getline( input, line ) ) {
+            line = trim( strip_slash( strip_comments( line ) ) );
+
+            //if( line.begin() == line.end() ) continue;
+
+            dst.append( line.begin(), line.end() );
+            dst.append( 1, '\n' );
+        }
+
+        return dst;
+    }
+
+    const std::string emptystr = "";
 
     struct ParserState {
+
+        struct file {
+            file( boost::filesystem::path p, const std::string& in ) :
+                path( p ), input( in )
+            {}
+
+            file( const std::string& in ) : path( "" ), input( in ) {}
+
+            boost::filesystem::path path;
+            string_view input;
+            size_t lineNR = 0;
+        };
+
         const ParseContext& parseContext;
         Deck * deck;
-        boost::filesystem::path dataFile;
+
         boost::filesystem::path rootPath;
-        /* the path map is shared between all state instances, as it has global
-         * resolution, i.e. PATHS specified in an included file can be used
-         * from the file that included it.
-         */
-        std::shared_ptr< std::map<std::string, std::string> > pathMap;
-        size_t lineNR;
-        std::shared_ptr<std::istream> inputstream;
+        std::map< std::string, std::string > pathMap;
+        std::list< std::string > input_strings;
+        std::list< file > input_files;
+        std::vector< file* > input_stack;
         RawKeywordPtr rawKeyword;
-        std::string nextKeyword;
+        string_view nextKeyword = emptystr;
 
+        boost::filesystem::path& current_path() {
+            return this->input_stack.back()->path;
+        }
 
-        ParserState(const ParserState& parent)
-            : parseContext( parent.parseContext )
-        {
-            deck = parent.deck;
-            pathMap = parent.pathMap;
-            rootPath = parent.rootPath;
+        const boost::filesystem::path& current_path() const {
+            return this->input_stack.back()->path;
         }
 
-        ParserState(const ParseContext& __parseContext)
-            : parseContext( __parseContext ),
-              deck( new Deck() ),
-              pathMap( std::make_shared< std::map< std::string, std::string > >() ),
-              lineNR( 0 )
-        {}
+        size_t& line() {
```

Comment:
It is: https://github.com/jokva/opm-parser/blob/internal-parser-procedure-use-ranges/opm/parser/eclipse/Parser/Parser.cpp#L622